src/task_behavior_ros/introspection.py

```diff
@@ -146,6 +153,60 @@ def publish_status(self):
         tree_status_msg.header.stamp = rospy.Time.now()
         self._status_pub.publish(tree_status_msg)
 
+    def publish_data_dump(self):
+        """
+        Publishes a data dump message of the tree
+        """
+        tree_data_dump_msg = self._create_tree_data_dump_msg(
+            self.parent, TreeDataDump())
+        tree_data_dump_msg.header.stamp = rospy.Time.now()
+        self._data_dump.publish(tree_data_dump_msg)
+
+    def _create_node_data_dump(self, node):
+        """ Create a node's data dump message
+        @param node [Node] The node to get the data from
+        @returns [NodeDataDump] The blackboard data from the node
+        """
+        nodedata = node._blackboard.get_memory(node._id)
+        ndd = NodeDataDump()
+        for key in nodedata.keys():
+            ndd.key.append(str(key))
+            ndd.value.append(str(nodedata.get_data(key)))
+        status = node.get_status()
+        if status.status is not NodeStatus.PENDING:
+            ndd.key.append('status')
+            ndd.value.append(str(status))
+        return ndd
+
+    def _create_tree_data_dump_msg(self, node, msg=TreeDataDump()):
+        """ Create the status message
+            @param node [Node] The node to get the status from (recursive)
+             @param msg [TreeDataDump] The message created so far
+            @returns [TreeDataDump] The status message
+        """
+        node_msg = TreeNode()
+        node_msg.id = str(node._id)
+        node_msg.name = node._name
+
+        data_msg = self._create_node_data_dump(node)
+        msg.data.append(data_msg)
+
+        if Behavior in type(node).__bases__:
+            node_msg.type = TreeNode.BEHAVIOR
+            for child in node._children:
+                node_msg.children.append(str(child._id))
```

Comment:
you can also create this list all at once node_msg.children = [str(child._id) for child in node._children]
it's usually more efficient than looping https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Loops